h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java

```diff
@@ -1,32 +1,163 @@
 package hex.tree.xgboost;
 
 import biz.k11i.xgboost.Predictor;
+import biz.k11i.xgboost.gbm.GBTree;
+import biz.k11i.xgboost.gbm.GradBooster;
+import biz.k11i.xgboost.tree.RegTree;
+import biz.k11i.xgboost.tree.RegTreeNode;
 import biz.k11i.xgboost.util.FVec;
 import hex.*;
 import hex.genmodel.GenModel;
+import hex.genmodel.algos.tree.SharedTreeGraph;
+import hex.genmodel.algos.tree.SharedTreeNode;
+import hex.genmodel.algos.tree.SharedTreeSubgraph;
 import hex.genmodel.algos.xgboost.XGBoostMojoModel;
 import hex.genmodel.algos.xgboost.XGBoostNativeMojoModel;
 import hex.genmodel.utils.DistributionFamily;
+import hex.tree.TreeBackedModel;
 import ml.dmlc.xgboost4j.java.*;
 import water.*;
 import water.fvec.Chunk;
 import water.fvec.Frame;
+import water.util.ArrayUtils;
 import water.util.Log;
 import hex.ModelMetrics;
 
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Objects;
 
 import static hex.tree.xgboost.XGBoost.makeDataInfo;
 import static hex.genmodel.algos.xgboost.XGBoostMojoModel.ObjectiveType;
 
-public class XGBoostModel extends Model<XGBoostModel, XGBoostModel.XGBoostParameters, XGBoostOutput> {
+public class XGBoostModel extends Model<XGBoostModel, XGBoostModel.XGBoostParameters, XGBoostOutput> implements TreeBackedModel {
 
   private XGBoostModelInfo model_info;
 
   public XGBoostModelInfo model_info() { return model_info; }
 
+  @Override
+  public SharedTreeGraph convert(final int treeNumber, final String treeClassName) {
+    GradBooster booster = null;
+    try {
+      booster = new Predictor(new ByteArrayInputStream(model_info._boosterBytes)).getBooster();
+    } catch (IOException e) {
+      Log.err(e);
+      H2O.fail(e.getMessage());
```

Comment:
never use H2O.fail when in h2o-algos, H2O.fail shuts down the cluster!

Reply:
Thank you, this is a leftover. Replaced with:

```java
    try {
      booster = new Predictor(new ByteArrayInputStream(model_info._boosterBytes)).getBooster();
    } catch (IOException e) {
      Log.err(e);
      throw new IllegalStateException("Booster bytes inaccessible. Not able to extract the predictor and construct tree graph.");
    }
```